### ISSUE TITLE
Automatically detect Install .app

### ISSUE THREAD

**Issue by n8felton** [MagerValp/AutoDMG#157]
As AutoDMG only allows you to build OS images equal to the version of the OS you are running AutoDMG on, it would be nice if on launch it could automatically detect and select the Install .app for the respective OS in the `/Applications` directory.

Full disclosure, this feature request comes from similar functionality available in http://diskmakerx.com

**Comment by MagerValp** [comment]
I agree that some kind of source auto detection could be convenient, but I tried DiskMaker X and I don't care for it at all. Apps should not throw up dialogs the first thing they do, it makes for a very hostile user experience.

Perhaps there's some other way of accomplishing it? Maybe automounting of the last used source, or a recent sources list in the File menu? It could be a bit of a minefield though, for example if the last used source was on a network drive and takes a minute to mount - the app is effectively locked while it's trying.

**Comment by n8felton** [comment]
I agree, the dialogs are annoying.

I was thinking more along the lines of the app automagically discovering the Installer .app for the booted version of the OS from /Applications on app launch and pre-populating the "Drop OS X Installer Here" with the Installer .app version it finds. Perhaps this could be cached/hashed in some way so that subsequent launches do not need to search and verify the Installer .app each time AutoDMG is launched. I'm not sure how quickly this new functionality would be at finding the .app, but just off the wall thoughts about trying to speed up the launch of AutoDMG if adding the search functionality were to delay launch times.

If someone doesn't want to use the one that is automagically detected from /Applications, they could still drag and drop a different installer.

Again, just some off the cuff thoughts to throw at the wall and see what sticks.

**Comment by erikng** [comment]
Not a bad idea, but if you're looking for more automation, perhaps you should use the CLI?

On Sep 21, 2016, at 7:43 AM, Nate Felton <notifications@github.com<mailto:notifications@github.com>> wrote:


I agree, the dialogs are annoying.

I was thinking more along the lines of the app automagically discovering the Installer .app for the booted version of the OS from /Applications on app launch and pre-populating the "Drop OS X Installer Here" with the Installer .app version it finds. Perhaps this could be cached/hashed in some way so that subsequent launches do not need to search and verify the Installer .app each time AutoDMG is launched. I'm not sure how quickly this new functionality would be at finding the .app, but just off the wall thoughts about trying to speed up the launch of AutoDMG if adding the search functionality were to delay launch times.

If someone doesn't want to use the one that is automagically detected from /Applications, they could still drag and drop a different installer.

Again, just some off the cuff thoughts to throw at the wall and see what sticks.

-

**Comment by n8felton** [comment]
I can agree with that, which I am already looking to do more with autodmg cache builder and some CI, this was just more of something that popped up because I only just discovered DiskMaker X last night and thought that feature to be worthy of potentially porting to AutoDMG.

**Comment by jazzace** [comment]
If a feature like this were to be added, I'd be tempted to make it a button much like the Download button for updates. Click on the button and it looks for "Install macOS Sierra" (or appropriate installer for OS) first in /Applications, then anywhere else on the system. If it finds it, it loads it just as if it had been dragged into the drop zone. If it doesn't find it, it could throw an error or it could simply present a standard Open dialog box.

I see this as the method that would be most consistent with the existing interface. I would not want it to automatically load, as AutoDMG does allow you to use cached disk images as well as installers, so it's not a given that the user will want to use an OS installer. (Very, very likely, but not a given.)

Anthony Reimer

**Comment by MagerValp** [comment]
I played with this a bit and settled on a new menu item: Select Installer… (hotkey ⌘I). It only shows (potentially) valid installers, and remembers the last directory you browsed, so ⌘I, I, Return will reload the last installer you selected, without using the mouse.

Please try [AutoDMG-1.7.1b481.dmg](https://www.dropbox.com/s/cipoxcikj74q4fu/AutoDMG-1.7.1b481.dmg) and let me know what you think.

**Comment by n8felton** [comment]
For me, the first time AutoDMG is run on a clean install, `⌘I` takes me to `~/Documents`, not `/Applications`. Not a huge issue, I can do `⌘+Shift+A` to navigate to `/Applications`, and `i` to select `Install OS X *`. Not sure if there is a way to make `/Applications` the default on the first run before subsequently remembering previously loaded installers/directories.

My current workflow often involves either re-imaging a test box to the OS I want to build an image for, or reverting a snapshot on a VM to a clean base install. In either case, there is almost never a previous run of AutoDMG or previously selected installer.

I'm still a fan of AutoDMG auto-discovering the appropriate `Install*.app` in `/Applications` based on the running version of the OS and pre-loading it without any user interaction needed; no dialog box or buttons to push. Perhaps then `⌘I` could be used to load a previously built image or to select a different installer than the one it auto-discovered. While not having any actual data to back up my theory, I'm assuming the majority of users who use the GUI for AutoDMG are following a similar workflow where they mostly only use the installer app found in `/Applications`.

Either way, this is a great addition and can still help speed up my workflow using the GUI until I have the time to automate more using the CLI. Regardless of whether auto-discovering becomes a thing, I hope `⌘I` make it to release.

**Comment by MagerValp** [comment]
Pre-selecting a file unfortunately clobbers the open panel's location saving feature, but I can make it default to the Applications folder. Try [AutoDMG-1.7.1b482](https://www.dropbox.com/s/crd7s84syxciwvt/AutoDMG-1.7.1b482.dmg).

You'll want to `defaults delete se.gu.it.AutoDMG NSNavLastRootDirectory` to clean out the old prefs, or test in a clean VM.

**Comment by MagerValp** [comment]
I just noticed that this now fights with the save dialog when clicking build, so I end up having to switch between my output folder and /Applications. Opening a template does the same. Less than ideal, I'm going to have to keep separate prefs for the three file panels.

**Comment by MagerValp** [comment]
Final draft of ⌘I here [AutoDMG-1.7.1b486](https://www.dropbox.com/s/vgunfrt9144ig5t/AutoDMG-1.7.1b487.dmg) (this build also contains the build success reporting from the other issue).

Automagic loading of the installer in /Applications I'm going to have to say no to. Best case it takes 3 seconds for my 15" with SSD with caches warm, in a VM or on slower hardware it takes a lot more. It'd drive me nuts if I had to wait that long every time I launched the app. There's also no good way of handling slow response or failures. I do agree that it matches the workflow of many users, but at the same time I want people to make a conscious decision on which installer they load.

I added ⌘⌥I hotkey though, that, for better or for worse, grabs the first installer it finds in /Applications.
